### DISCUSSION TITLE
How should a metaclass __call__ be annotated to satisfy pyright

### DISCUSSION THREAD

**Discussion by NeilGirdhar** [microsoft/pyright#5561] [Q&A] [answered]
I'm getting some errors with Pyright and [Haiku](https://github.com/deepmind/dm-haiku).  Here's a MWE fo the error:
```python
from typing import TypeVar, Protocol

T = TypeVar("T")

class ModuleMetaclass(type(Protocol)):
    def __call__(self: type[T], *args, **kwargs) -> T:
        return super().__call__(self, *args, **kwargs)

class Module(metaclass=ModuleMetaclass):
    pass

x = Module()
```
gives
```
/home/neil/src/dm-haiku/a.py
  /home/neil/src/dm-haiku/a.py:12:5 - error: Could not bind method "__call__" because "ModuleMetaclass" is not assignable to parameter "self"
    Type "Module" cannot be assigned to type "type[T@__call__]"
```
I can make the errors go away by changing it to:
```python
    def __call__(self: ModuleMetaclass, *args, **kwargs) -> Any:
```
Is this recommended?  Ideally, if intersections were added, would the annotation be:
```python
    def __call__(self: type[T] & ModuleMetaclass, *args, **kwargs) -> T:
```
Am I missing something?

**Comment by erictraut** [comment, marked as answer]
Both mypy and pyright generate the same error for the code above because `T` is not a subtype of `ModuleMetaclass`. The bound you've provided for the TypeVar is incorrect. There is no way to express that `T` must satisfy the constraint that "the class that is instantiated to create T is a subtype of `ModuleMetaclass`".

To fix this, simply remove the bound. The type error will then go away — at least in pyright. With mypy, you then run into a bug in where it doesn't properly handle the `super()` call within a metaclass, and it reports a false positive error.

Note that if you use a custom `__call__` method in your metaclass, pyright currently makes the assumption that your metaclass may be behaving in a way that differs from `type.__call__`, so it doesn't type check the arguments against your class' `__new__` or `__init__` methods. Refer to [this discussion](https://github.com/microsoft/pyright/issues/5488) for more details if you're interested.

**Reply by NeilGirdhar** [reply]
I need to investigate this further because pyright is giving errors with no bound with the Haiku metaclass `ModuleMetaclass`.

**Reply by NeilGirdhar** [reply]
I produced a new MWE that produces errors in Pyright.  Would you mind taking a look when you have a chance?

**Reply by erictraut** [reply]
In your updated code, you're calling `type` and passing it a special form (`Protocol`). This may work at runtime, but a static type checker isn't going to know what to do with it. Replace `type(Protocol)` with `type`, and the type errors will go away.

**Reply by NeilGirdhar** [reply]
Thank you.  (It's not my code.  It's from here: https://github.com/deepmind/dm-haiku.  I'm just trying to figure how to repair the errors it's giving me.)     I'll do it your way then.  Thanks again.

**Reply by NeilGirdhar** [reply]
They have a comment saying
```
# We subclass `type(Protocol)` in order to avoid metaclass conflicts when
# defining modules that also inherit from `Protocol`. Note that `type(Protocol)`
# already inherits from `abc.ABCMeta`.
```
So I guess I probably can't change that.
